### PR TITLE
Implement host list.

### PR DESCRIPTION
**Pull request by dmtucker**
We will need to modify `config.request` to accommodate POST and PUT requests soon.

### CODE REVIEW

File: yaas/host.py [hunk 1]
```diff
@@ -1,4 +1,30 @@
...
+    for name, fn in subcommands.items():
+        subparsers.add_parser(name, help=inspect.getdoc(fn))
+    subargs, extra = parser.parse_known_args(args)
+    subcommands[subargs.subcommand](
+        subparsers.choices[subargs.subcommand],
+        extra)
+
+
+def host_list(parser, args):
+    """ List all server agents. """
```
Comment (karth295) [inline yaas/host.py]: "server agents" is a weird term. Maybe "List all registered hosts"?

File: yaas/__main__.py [hunk 2]
```diff
@@ -34,13 +36,19 @@ def main():
...
         version=YAAS_VERSION,
-        help="print the yaas version")
+        help=inspect.getdoc(version))
 
     parser.add_argument(
         '--verbose',
         action='store_true',
         default=False,
-        help="print ambari api requests and responses")
+        help="Print the raw Ambari response.")
```
Comment (karth295) [inline yaas/__main__.py:48]: So should this be --raw?

File: yaas/config.py [hunk 3]
```diff
@@ -3,28 +3,39 @@
...
-        }
-
-    if args.verbose:
-        opts['hooks'] = {
-            'response': print_request_and_response,
-            }
-
-    return opts
 
+def request(endpoint, callback):
```
Comment (karth295) [inline yaas/config.py]: "callback" makes it sound like this is an asynchronous request. Maybe print_function?

File: yaas/host.py [hunk 4]
```diff
@@ -1,4 +1,30 @@
...
+from __future__ import print_function
+
+import inspect
+
+from . import config
+
+
 def command(parser, args):
-    pass
+    """ Interact with server agents. """
```
Comment (karth295) [inline yaas/host.py]: "server agents" is weird wording. Maybe "view and update registered hosts"?

File: yaas/__main__.py (hunk 2 above)
Comment (dmtucker) [inline yaas/__main__.py:48]: oo baby I like it `--raw`.

File: yaas/config.py (hunk 3 above)
Comment (dmtucker) [inline yaas/config.py]: good point

**Comment by dmtucker** [comment]
`yaas host list --fields=Hosts/disk_info` output could be better.

File: yaas/host.py [hunk 5]
```diff
@@ -1,4 +1,56 @@
...
+    for name, fn in subcommands.items():
+        subparsers.add_parser(name, help=inspect.getdoc(fn))
+    subargs, extra = parser.parse_known_args(args)
+    subcommands[subargs.subcommand](
+        subparsers.choices[subargs.subcommand],
+        extra)
+
+
+def _list(parser, args):
+    """ List all registered agents. """
```
Comment (karth295) [inline yaas/host.py]: This should be registered hosts. If you use the server to deploy agents, it's almost an implementation detail that the server uses agents. The endpoint is called hosts, not agents.

File: yaas/host.py [hunk 6]
```diff
@@ -1,4 +1,56 @@
...
+    try:
+        response.raise_for_status()
+    except requests.exceptions.HTTPError:
+        print(response.json()['message'], file=sys.stderr)
+        sys.exit(1)
+    for item in response.json()['items']:
+        line = []
+        line_keys = ['host_name', 'cluster_name']
+        for key in line_keys:
+            if item['Hosts'].get(key) is not None:
```
Comment (karth295) [inline yaas/host.py]: Why mix [] notation and .get()? I'd vote we stick to [].

File: yaas/host.py [hunk 7]
```diff
@@ -1,4 +1,56 @@
...
+                line.append(item['Hosts'][key])
+        print(' - '.join(line))
+        for key, value in item['Hosts'].items():
+            if key not in line_keys:
+                print('{indent}{key}: {value}'.format(
+                    indent=' '*4, key=key, value=value))
```
Comment (karth295) [inline yaas/host.py]: We shouldn't print extra information. Users should use --raw if they need information we don't consider important.

File: yaas/host.py (hunk 6 above)
Comment (dmtucker) [inline yaas/host.py]: There's no guarantee that a given `key` will be set (e.g. a host not in a cluster will not have a `cluster_name`). Just using `item['Hosts'][key]` could result in a `KeyError`.

File: yaas/host.py (hunk 7 above)
Comment (dmtucker) [inline yaas/host.py]: `--fields` can only be used with `yaas --raw` then? I'm meh about either solution... Printing random key/value pairs isn't great, but not supporting `--fields` on normal output isn't great either.

**Comment by dmtucker** [comment]
```
$ yaas host list --fields=Hosts/host_state,Hosts/host_status
host1.example.com - cluster1
    host status: ALERT
    host state: HEALTHY
host2.example.com - cluster1
    host status: HEALTHY
    host state: HEALTHY
host3.example.com - cluster1
    host status: HEALTHY
    host state: HEALTHY
host4.example.com - cluster1
    host status: HEALTHY
    host state: HEALTHY
host5.example.com - cluster1
    host status: ALERT
    host state: HEALTHY
host6.example.com - cluster1
    host status: HEALTHY
    host state: HEALTHY
```

File: yaas/host.py (hunk 6 above)
Comment (karth295) [inline yaas/host.py]: if key in item['Hosts']

**Pull request merged**